### ISSUE TITLE
"go build" doesn't parallelize compilations within a package

### ISSUE THREAD

**Issue by petermattis** [golang/go#9887]
The lack of parallelization for compilations within a single package doesn't matter too much when building go files, but does affect cgo compilation. I have a package with about 32k lines of (generated) C++. Compiling it serially takes ~10.5s. A small hack to cmd/go/build.go to compile C/C++/Obj-C files in parallel brings this down to ~3.7s.

The patch I hacked together to parallelize cgo compilation work is ~50 lines of deltas and is mostly contained in `builder.cgo()`.

PS The compilation times above were measured using go 1.4.1 on Mac OS X, but this code hasn't changed at head and similar compilation times can be seen there.

**Comment by robpike** [comment]
Seems reasonable to me. Why not send the patch?

**Comment by petermattis** [comment]
Done: https://go-review.googlesource.com/#/c/4931/

**Comment by petermattis** [comment]
I'm curious if there is any plan to parallelize cgo compilation for go1.7. The finer-grained work scheduling for go tool actions mentioned in https://go-review.googlesource.com/#/c/4931/ would be great if it ever happens, though it doesn't look like there is any movement on that front.

**Comment by bcmills** [comment]
Contrast #21605. Are there any similar ordering issues for other languages?

**Comment by ianlancetaylor** [comment]
Currently we only support C, C++, and Fortran, and C and C++ have no ordering issues.

I'm not aware of ordering issues for any other language.

**Comment by r0l1** [comment]
Any updates on this? Compiling one of our packages takes around 30 seconds...

**Comment by bcmills** [comment]
@r0l1, this issue is milestoned to `Unplanned`, meaning that we do not currently plan to work on it.

The previous CL was closed in favor of “implementation of finer-grained actions”, which I think may be #8893.

**Comment by mvdan** [comment]
I had a brief look at this last week, because this is causing an annoying build time bottleneck in a project of mine. I assume the file we're interested in is `src/cmd/go/internal/work/exec.go`.

In particular, the `Builder.cgo` method makes calls to `Builder.gcc` synchronously in a loop.

It's unclear to me how we'd fix this, though. Can anyone point me to a doc that explains how `work.Action` is designed at a high level?

For example, I can imagine these ways to fix the issue, but it's unclear which would be the right one:

* In `Builder.cgo`, create a child `Action` for each `gcc` call, and ensure they're all run before we continue (is this possible?)
* Before the entire `cgo` method is called via `build`, separate the work into a tree of actions (how? `cfiles` is constructed as part of `cgo`, for example)

**Comment by mvdan** [comment]
I had a sudden realisation: one of the packages that are really slow to compile with cgo, https://github.com/olebedev/go-duktape/tree/v3, simply bundles a ton of C code as a [single 3.5MiB file](https://github.com/olebedev/go-duktape/blob/v3/duktape.c). I presume that parallelizing cgo compilation per-file wouldn't help at all in that case, and I also assume that there is no way to make that any faster.

Luckily, that's not the norm. For example, https://github.com/ipsn/go-libtor builds tor from its C source, directly building its thousands of C files. That seems like it would get a big win from a fix for this issue.

**Comment by diamondburned** [comment]
> I had a brief look at this last week, because this is causing an annoying build time bottleneck in a project of mine. I assume the file we're interested in is `src/cmd/go/internal/work/exec.go`.
> 
> In particular, the `Builder.cgo` method makes calls to `Builder.gcc` synchronously in a loop.

From my observation, both `cmd/go/internal/work` and `cmd/cgo` invoke `gcc`
sequentially a lot of times, but `cmd/cgo` takes up the bulk of the time.

I was able to write a prototype that attempts to parallelize both of them,
however the CL for `cmd/go/internal/work` was denied in favor of another
unplanned refactor. I still believe that parallelizing just `cmd/cgo` will
significantly improve compile time, though.

If anybody wants to test how much faster the changes are, it can be found at
[diamondburned/go/work-gcc-parallelism](https://github.com/diamondburned/go/tree/work-gcc-parallelism).
When using this branch, use `./make.bash` to compile instead, as the change
currently fails a Cgo test that I'm still not sure why.

I'm currently daily driving Go 1.17 with these 2 patches applied on top almost
daily, and I find compile time to be roughly 3-4x faster. The implementation
will limit the maximum workers at 4 per package regardless of the thread count,
so the number is to be expected.

The branch doesn't account for Fortran, however, so it might not work at all for
codebases that need those.

**Comment by r0l1** [comment]
@diamondburned thanks for the info. I am currently testing your patch (`cmd/go/internal/work`) on latest go master and all tests pass. Could you tell me more about the first patch (`cmd/cgo`)? I didn't find it. Was it already merged into master?

**Comment by diamondburned** [comment]
Just as a small update, branch [go-cgo-gcc-parallelism-1.18](https://github.com/diamondburned/go/tree/go-cgo-gcc-parallelism-1.18) now tracks the 2 patches for Go 1.18.

For Nix users, the drop-in overlay is

```nix
(self: super: {
	go = super.go.overrideAttrs (old: {
		version = "1.18";
		src = builtins.fetchurl {
			url    = "https://golang.org/dl/go1.18.linux-amd64.tar.gz";
			sha256 = "0kr6h1ddaazibxfkmw7b7jqyqhskvzjyc2c4zr8b3kapizlphlp8";
		};
		doCheck = false;
		patches = [
			# cmd/go/internal/work: concurrent ccompile routines
			(builtins.fetchurl "https://github.com/diamondburned/go/commit/ec3e1c9471f170187b6a7c83ab0364253f895c28.patch")
			# cmd/cgo: concurrent file generation
			(builtins.fetchurl "https://github.com/diamondburned/go/commit/50e04befeca9ae63296a73c8d5d2870b904971b4.patch")
		];
	});
})
```

**Comment by Evengard** [comment]
@diamondburned Unfortunately your code have problems. When attempting to compile my project with your version, it failed with a rather cryptic error:
```
cgo: duplicate pkg.Name nox_xxx_objGetTeamByNetCode_418C80, &{nox_xxx_objGetTeamByNetCode_418C80 _Cfunc_nox_xxx_objGetTeamByNetCode_418C80 nox_xxx_objGetTeamByNetCode_418C80  func <nil> 0xc001b12870 false } == &{nox_xxx_objGetTeamByNetCode_418C80 _Cfunc_nox_xxx_objGetTeamByNetCode_418C80 nox_xxx_objGetTeamByNetCode_418C80  func <nil> 0xc0007171d0 false }
```
The project compiles fine without multithreaded cgo though (aka regular Golang build)
The project in question is https://github.com/noxworld-dev/opennox, if need to check the bug out.

**Comment by diamondburned** [comment]
> Unfortunately your code have problems.

The modifications are definitely band-aid. It works for my purposes so I've stopped there, because `cmd/cgo`'s code is very complicated and weird for me to work with.

If you can't get the compiler to work with just the `cmd/cgo` patch, then you might need to investigate deeper yourself.

**Comment by diamondburned** [comment]
I've made [go-cgo-gcc-parallelism-1.19](https://github.com/diamondburned/go/tree/go-cgo-gcc-parallelism-1.19) to track Go 1.19.

**Comment by Evengard** [comment]
https://github.com/Evengard/go/tree/go-cgo-gcc-parallelism-1.20.3 tracks the Go 1.20.3 release for anyone who needs that.

**Comment by podtserkovskiy** [comment]
It's been a while since this issue was created and two CLs were abandoned in favor to properly redesign `cmd/internal/work` to make actions API more fine grained. 

This is definetely rught thing to do, but nobody had time to implement that.
I do anticipate that this is not an easy project and personally I don't have enough time to invest into it at the moment.

Are we happy to do a tiny fix to just parallelize the work inside `Builder.cgo` as a temporary solution?

**Comment by podtserkovskiy** [comment]
I've tried to keep it as less invasive as possible, let me know if it works for you
https://go-review.googlesource.com/c/go/+/579815

**Comment by matloob** [comment]
It would be ideal to try to fix this by making actions more fine grained, especially since there aren't any other cases where we do work concurrently within an action. It complicates to have two layers of concurrency- one layer being the actions that are running and the other being the C compilations within the action.

**Comment by diamondburned** [comment]
> It would be ideal to try to fix this by making actions more fine grained, especially since there aren't any other cases where we do work concurrently within an action. It complicates to have two layers of concurrency- one layer being the actions that are running and the other being the C compilations within the action.

AFAIK, this was one of the reasons for not merging the CLs. I definitely agree that that's the ideal solution, but as @podtserkovskiy pointed out, it wasn't really being worked on, unfortunately.

**Comment by podtserkovskiy** [comment]
> It would be ideal to try to fix this by making actions more fine grained

I fully agree that the redesign of the actions API would be ideal, but this is not a trivial change and it requires much more time for proper implementation than this fix.

> there aren't any other cases where we do work concurrently within an action

I can recall one more place with a potential to execute things concurrently [`gcToolchain.asm`](https://github.com/golang/go/blob/2073b35e07ce9cea47ee1fbe763b304d2371954f/src/cmd/go/internal/work/gc.go#L397-L405) within an action.

**Comment by matloob** [comment]
> > there aren't any other cases where we do work concurrently within an action
> 
> I can recall one more place with a potential to execute things concurrently [`gcToolchain.asm`](https://github.com/golang/go/blob/2073b35e07ce9cea47ee1fbe763b304d2371954f/src/cmd/go/internal/work/gc.go#L397-L405) within an action.

Sorry, I meant that there aren't any other places currently where we do work concurrently with an action

**Comment by diamondburned** [comment]
Yeah, unfortunately, `cmd/go/internal/work` is only the slow one in some cases. Most of the bottleneck is in `cmd/cgo` which is its own beast.

**Comment by podtserkovskiy** [comment]
I've published parallelisation of [`gccDefines`](https://github.com/golang/go/blob/960fa9bf66139e535d89934f56ae20a0e679e203/src/cmd/cgo/gcc.go#L2192) calls in `cmd/cgo`.

There are more `CC` calls to make concurrent.  

But even this change makes processing of "net" package (5 CGo files) 25% faster.
it won't improve performance of packages with a single CGo file, the more CGo files we have, the higher the performance increase.

CL: https://go-review.googlesource.com/c/go/+/581336

**Comment by mappu** [comment]
Hi, I have [a very large CGO module](https://github.com/mappu/miqt). The /qt/ package alone has 340 .cpp files plus 340 .go files using CGO.

Since CL 581336 was merged, I tried to compile the package using `gotip` (go version devel go1.24-e67c0f0), on 4-core Neoverse N1. But, there was no noticeable speedup versus go1.23.3 nor versus old go1.19.8.

**Comment by leaanthony** [comment]
Hello from the future. Is there any plans for this to go ahead? The main pain point here is the need to use CGO on non-Windows machines for some large libraries (QT & GTK). These take a **ridiculous** amount of time to compile. To be clear, I'd *much* prefer we had something like purego in the standard library, but if that's not going to happen, parallel CGO compiles would go an awful long way to helping some serious pain points. Thanks all 🙏

**Comment by matloob** [comment]
Hi, we're trying to determine whether to prioritize the work to fix this properly (with fine grained actions) for Go 1.26. Please let me know if your workflow is affected by this, and if you'd be able to test out the changes to see if it improves your workflow.

Thanks!

**Comment by ilius** [comment]
@matloob [miqt](https://github.com/mappu/miqt) library is a good example, because it takes about 10 minutes to compile on an intel `i5-12400`.
I can test it on my [ayandict](https://github.com/ilius/ayandict) app.

**Comment by DeedleFake** [comment]
@matloob, my workflow on Trayscale is affected by this. Compilations without the cache can take 20 to 30 minutes, which makes testing small changes incredibly annoying. I don't know how much something like this would help, but if it could at all that would be very much appreciated.

And yes, I'd be able to test it.

**Comment by matloob** [comment]
@ilius @DeedleFake Hello! We have a CL that's ready to test! Check out https://golang.org/cl/699435

It includes fixes for this issue as well as some speedups in the cgo command (#75167)

You can install `golang.org/dl/gotip@latest` and then run `gotip download 699435` to build it and then run `gotip` as your go command.

There's still some more work that can be done, but these changes will deliver most of the benefit of the work. Let me know if you have any feedback!

Also if you're at Gophercon please come and say hi!

**Comment by ilius** [comment]
@matloob It seems to use all CPU cores, and takes about 3 minutes instead of 10 minutes.
Now I wish for a flag / env var to limit CPU usage (like `%400` meaning 4 cores) because my PC can not handle full CPU usage and gets dangerously hot!

**Comment by rcalixte** [comment]
> Now I wish for a flag / env var to limit CPU usage (like `%400` meaning 4 cores) because my PC can not handle full CPU usage and gets dangerously hot!

Many other build systems use something like `-j` for this. I'm not sure if it makes sense here or how feasible it would be to implement but just in case anyone is wondering about prior art.

**Comment by matloob** [comment]
@ilius @rcalixte You could use the `-p` flag to go build,  which is set by default to `GOMAXPROCS`. That will restrict the number of compile invocations that happen at the same time. So you could try `go build -p 4` or `GOMAXPROCS=4 go build`.

**Comment by gucio321** [comment]
Also, it seems to ignore some `#cgo` commands (or something has been updated?)

We have the following: `// #cgo linux CXXFLAGS: -Wno-changes-meaning -Wno-invalid-conversion -fpermissive` and with this change I'm getting this:
```console
# github.com/AllenDang/cimgui-go/imnodes
In file included from wrapper.h:7,
                 from wrapper.cpp:4:
../cwrappers/cimnodes.h:111:29: error: declaration of ‘EmulateThreeButtonMouse ImNodesIO::EmulateThreeButtonMouse’ changes meaning of ‘EmulateThreeButtonMouse’ [-Wchanges-meaning]
  111 |     EmulateThreeButtonMouse EmulateThreeButtonMouse;
      |                             ^~~~~~~~~~~~~~~~~~~~~~~
../cwrappers/cimnodes.h:111:5: note: used here to mean ‘typedef struct EmulateThreeButtonMouse EmulateThreeButtonMouse’
  111 |     EmulateThreeButtonMouse EmulateThreeButtonMouse;
      |     ^~~~~~~~~~~~~~~~~~~~~~~

```

**Comment by matloob** [comment]
@gucio321 Thanks for the report! If the build is open source, could you give me reproduction instructions?

**Comment by gucio321** [comment]
@matloob sure! Its https://github.com/AllenDang/cimgui-go just build the examples/glfw

I'm on fedora 42 btw.

**Comment by mappu** [comment]
Thank you @matloob !

On [CL 699435](https://go.dev/cl/699435) (a072fe73c2) I also get some warnings building github.com/mappu/miqt/examples/helloworld (Qt 5): `cc1plus: warning: command-line option ‘-std=gnu11’ is valid for C/ObjC but not for C++`. Is it perhaps mixing up the CFLAGS and CXXFLAGS? This flag is being set in a `#cgo` comment.

Although they were only warnings. The build succeeded and is significantly faster on my PC:

||go1.24.4|gotip for [CL 699435](https://go.dev/cl/699435)<br>go1.26-devel_a072fe73c2)|CL + taskset †
|---|---|---|---
|Debian 13<br>6c12t Ryzen 7600<br>`time go build`|<pre>real    6m22.398s<br>user    6m57.211s<br>sys     0m28.964s</pre>|<pre>real    1m41.935s<br>user    11m20.216s<br>sys     0m43.845s</pre>|<pre>real    1m45.823s<br>user    7m29.765s<br>sys     0m28.418s</pre>

***

† I was concerned that the `user` time had increased a lot in this CL. I assumed it was because the build uses all logical cores (SMT) to no benefit. So I limited it to only one in each core pair with `taskset -c 0,1,2,3,4,5`. This had almost no effect on the `real` time but brought the `user` time back roughly in line with go1.24.

**Comment by matloob** [comment]
@gucio @mappu The latest patchset should fix the issue (we indeed were passing the cflags instead of the c++ for the c++ builds). I've tested it with both of your examples.

**Comment by leaanthony** [comment]
Nothing to report but just wanted to thank you for doing this as it will make a huge impact 🙏

**Comment by DeedleFake** [comment]
Finally got a chance to test this on the particular project that I've been dealing with long compilations on.

Edit: This is on an AMD Ryzen 5 7640U. That's a 6 core CPU.

### go version go1.25.0 linux/amd64
```
Executed in  324.22 secs    fish           external
   usr time  493.88 secs    0.00 micros  493.88 secs
   sys time   59.28 secs  977.00 micros   59.28 secs
```

### go version go1.26-devel_8520b5bbf Mon Sep 1 11:14:52 2025 -0400 X:arenas,synctest,jsonv2,greenteagc linux/amd64
```
Executed in  108.79 secs    fish           external
   usr time  783.04 secs    0.00 micros  783.04 secs
   sys time   84.53 secs  920.00 micros   84.53 secs
```

### go version go1.26-devel_8520b5bbf Mon Sep 1 11:14:52 2025 -0400 linux/amd64
```
Executed in  108.14 secs    fish           external
   usr time  779.76 secs    0.00 micros  779.76 secs
   sys time   84.51 secs  840.00 micros   84.51 secs
```

Pretty definite improvement. Very nice job.
